src/ivc/incrementally_verifiable_computation.rs

```diff
@@ -228,17 +236,84 @@ where
 
     pub fn verify<const T: usize, RP1, RP2>(
         &mut self,
-        _pp: &PublicParams<'_, A1, A2, T, C1, C2, SC1, SC2, RP1, RP2>,
-        _steps_count: usize,
-        _z0_primary: [C1::Scalar; A1],
-        _z0_secondary: [C2::Scalar; A2],
+        pp: &PublicParams<'_, A1, A2, T, C1, C2, SC1, SC2, RP1, RP2>,
+        num_steps: NonZeroUsize,
+        primary_z_0: [C1::Scalar; A1],
+        secondary_z_0: [C2::Scalar; A2],
     ) -> Result<(), Error>
     where
-        RP1: ROPair<C1::Scalar>,
-        RP2: ROPair<C2::Scalar>,
+        RP1: ROPair<C1::Scalar, Config = MainGateConfig<T>>,
+        RP2: ROPair<C2::Scalar, Config = MainGateConfig<T>>,
     {
-        // TODO #31
-        todo!("Logic at `RecursiveSNARK::verify`")
+        if num_steps.get() != self.step {
+            return Err(Error::NumStepNotMatch);
+        }
+        if self.primary.z_0 != primary_z_0 || self.secondary.z_0 != secondary_z_0 {
+            return Err(Error::InputHashNotMatch);
+        }
+
+        // verify X0
+        let ro1 = &mut RP1::OffCircuit::new(pp.primary.params.ro_constant.clone());
+        // TODO: cache the result in IVC struct (?) so that we don't need to re-calculate it
```

Comment:
TODO only with issue number, here you can use my draft with `prove_step` - #164

Reply:
I will remove TODO here because I found the recalculation in multiple places. I will work on a new PR for the digest method.